### PR TITLE
lib: Factor out getting IP address from net.Addr

### PR DESCRIPTION
**Pull request by greatroar**
... and add fast paths for common cases.

### CODE REVIEW

File: lib/pmp/pmp.go [hunk 1]
```diff
@@ -66,10 +67,8 @@ func Discover(ctx context.Context, renewal, timeout time.Duration) []nat.Device
 	conn, err := (&net.Dialer{}).DialContext(timeoutCtx, "udp", net.JoinHostPort(ip.String(), "5351"))
 	if err == nil {
 		conn.Close()
-		localIPAddress, _, err := net.SplitHostPort(conn.LocalAddr().String())
-		if err == nil {
-			localIP = net.ParseIP(localIPAddress)
-		} else {
+		localIP, _ = osutil.IPFromAddr(conn.LocalAddr())
+		if localIP == nil {
```
Comment (calmh) [inline lib/pmp/pmp.go:71]: err is nil in the line below, so some tweak required here to the error handling

**Review by calmh — COMMENTED** [review]
Looks good

File: lib/pmp/pmp.go (hunk 1 above)
Comment (greatroar) [reply lib/pmp/pmp.go:71]: This function doesn't return an error.

Comment (greatroar) [reply lib/pmp/pmp.go:71]: Is that why DeepSource is failing? It's giving me a 500.

Comment (calmh) [reply lib/pmp/pmp.go:71]: It may not return an error, but we're logging an error which is guaranteed to be nil

Comment (greatroar) [reply lib/pmp/pmp.go:71]: Oh, right! Fixed.

**Review by calmh — APPROVED** [review]

**Pull request merged**